typescript/baseDAO/src/api.ts

```diff
@@ -0,0 +1,230 @@
+// SPDX-FileCopyrightText: 2020 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+import { ContractMethod , Signer, TransactionWalletOperation,
+         Contract, ContractAbstraction,
+         ContractProvider, TezosToolkit, MichelsonMap
+       } from '@taquito/taquito';
+import { InMemorySigner, importKey } from '@taquito/signer';
+
+import { Parameter } from './generated/Parameter';
+import { Startup } from './generated/Startup';
+import { CallCustom } from './generated/CallCustom';
+import { Accept_ownership } from './generated/Accept_ownership';
+import { Burn } from './generated/Burn';
+import { Balance_of } from './generated/Balance_of';
+import { Confirm_migration } from './generated/Confirm_migration';
+import { Drop_proposal } from './generated/Drop_proposal';
+import { Flush } from './generated/Flush';
+import { Get_total_supply } from './generated/Get_total_supply';
+import { GetVotePermitCounter } from './generated/GetVotePermitCounter';
+import { Migrate } from './generated/Migrate';
+import { Mint } from './generated/Mint';
+import { Propose } from './generated/Propose';
+import { Set_quorum_threshold } from './generated/Set_quorum_threshold';
+import { Set_voting_period } from './generated/Set_voting_period';
+import { Transfer } from './generated/Transfer';
+import { Transfer_contract_tokens } from './generated/Transfer_contract_tokens';
+import { Transfer_ownership } from './generated/Transfer_ownership';
+import { Update_operators } from './generated/Update_operators';
+import { Vote } from './generated/Vote';
+
+function println(s: string) {
+  console.log(s);
+}
+
+type Callback = (Contract) => ContractMethod<ContractProvider>
+
+type ExtractMapKey<MapType> = MapType extends Map<infer R, infer V> ? R : null;
+type ExtractMapValue<MapType> = MapType extends Map<infer R, infer V> ? V : null;
+
+const unit = ["Unit"]; //https://github.com/ecadlabs/taquito/issues/526
+
+/**
+ * Wraps BaseDAO Contract methods
+ */
+export class BaseDAOContract {
+  private contractAddr: string;
+  private nodeAddr: string;
+  private senderSk: string;
+  private contract: undefined | Promise<ContractAbstraction<ContractProvider>>;
+
+  public debug: boolean;
+
+  public lastOperationHash: undefined | string;
+
+  constructor(nodeAddr: string, senderSk: string, contractAddr: string) {
+    this.contractAddr = contractAddr;
+    this.nodeAddr = nodeAddr;
+    this.senderSk = senderSk;
+    this.debug = false;
+  }
+
+  private initContract(): Promise<ContractAbstraction<ContractProvider>> {
+    if (this.contract === undefined) {
+      const Tezos = new TezosToolkit(this.nodeAddr);
+      Tezos.setProvider({ signer: new InMemorySigner(this.senderSk) });
+      this.contract = Tezos.contract.at(this.contractAddr);
+      return this.contract;
+    } else return this.contract;
+  }
```

Comment:
Both branches of this method return `this.contract`. I think you may remove the `else` branch and have `return this.contract` just after the `if` statement.

Reply:
I thought TS is just not smart enough to figure that `this.contract` is not undefined just after the `if` statement, unless it is in a type refining branch of an if/else statement. Trying it anyway..

Reply:
Apparently it works, but I don't understand how. Because this works as well

```
  private initContract(): Promise<ContractAbstraction<ContractProvider>> {
      return this.contract;
  }
```

Why does it allow unification of `undefined | a` with `a` ?

Reply:
I think it needs to be compiled with `strictNullCheck`, or just the strict mode, from what I've read. Apart from this, I'm not sure why such a thing is allowed...

Reply:
Yep, just tried it on https://www.typescriptlang.org/play
If you have this code:
```typescript
class Foo {
    private x: number | undefined
    public foo(): number {
        return this.x
    }
}
```

It will accuse an error, unless you disable `strictNullChecks`.

Reply:
Even after enabling the strict null checks, your suggestion works for me. So it turns out that the Typescript compiler is actually smarter than I imagined, and can actually infer that `this.contract` is not null after the non-null assignment in the `if` block. 

That is reeally impressive.

Reply:
Have included this change now.